cpp/src/parquet/metadata.cc

```diff
@@ -793,26 +793,29 @@ class FileMetaData::FileMetaDataImpl {
 
     std::shared_ptr<FileMetaData> out(new FileMetaData());
     out->impl_ = std::make_unique<FileMetaDataImpl>();
-    out->impl_->metadata_ = std::make_unique<format::FileMetaData>();
-
-    auto metadata = out->impl_->metadata_.get();
-    metadata->version = metadata_->version;
-    metadata->schema = metadata_->schema;
+    {
+      // GH-39336: Copy the metadata, but only the row groups we need.
+      auto temp_row_groups = std::move(this->metadata_->row_groups);
+      try {
```

Comment:
Why do you add try-catch block?

Reply:
I think `make_unique` and copying might throw exception( like `std::bad_alloc`), and even when ex is thrown, we'd better keep `this` not changed (otherwise it would be hard to debugging?)

Reply:
I'm not sure if it worths this kind of complexity to handle the exception just in order to avoid the copy.